### PR TITLE
JS: only calculate getStringValue for concatenation roots

### PR DESCRIPTION
**Pull request by erik-krogh**
After https://github.com/github/codeql/pull/3752 I thought about computing `getStringValue()` differently. 

I tried to use `strictconcat` to compute `getStringValue()` for `AddExpr`.   
In terms of performance, this only makes sense if `getStringValue()` is only defined for the "concatenation-roots".   
And I had to factor out all the `getStringValue()` overrides in order to avoid non-monotonic recursion.  

This also means that there are [lots of `AddExpr` where `getStringValue()` is no longer defined](https://lgtm.com/query/6879957369687162983/).  
But they are all themselves contained in another concatenation.   
If we ignore the `AddExpr` that are themselves contained in a string concatenation, then there is [no change in results](https://lgtm.com/query/7711412374264571780/). 

I don't know if this is a good idea, or if we will loose any results. 

[Performance is very good](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1593615175727), and almost perfectly cancels the performance regressions we had [since rc/1.24](https://git.semmle.com/asger/dist-compare-reports/tree/js/without-new-queries_1593275994181).  
(I have no idea if the two are related).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Expr.qll [hunk 1]
```diff
@@ -1517,13 +1513,64 @@ class URShiftExpr extends @urshiftexpr, BinaryExpr {
...
+}
+
+/**
+ * Gets the concatenated string for a string-concatenation `add`.
+ * Only has a result if `add` is not itself an operand in another string-concatenation.
+ */
+private string getConcatenatedString(Expr add) {
+  result = getConcatenatedString(add.getUnderlyingValue())
+  or
+  not add = getAnAddOperand(_) and
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/Expr.qll:1555]: Don't we lose the following concatenation (permute the branches as needed)?

```
unknown + "foo" + "bar"
//        ^^^^^^^^^^^^ should be "foobar"

```

If we don't mind that loss (I think we do mind). Shouldn't we use `getUnderlyingValue` here as well?

File: javascript/ql/src/semmle/javascript/Expr.qll [hunk 2]
```diff
@@ -1517,13 +1513,64 @@ class URShiftExpr extends @urshiftexpr, BinaryExpr {
...
+  forex(Expr leaf | leaf = getAnAddOperand*(add) and not exists(getAnAddOperand(leaf)) |
+    exists(getConstantString(leaf))
+  ) and
+  result =
+    concat(Expr leaf |
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/Expr.qll:1560]: `strictConcat` is implied by the previous conjunct. I think the optimizer would prefer to know that.

Also see `ConcatenationRoot::getConstantStringParts`, if you haven't already.

File: javascript/ql/src/semmle/javascript/Expr.qll (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/Expr.qll:1555]: > If we don't mind that loss (I think we do mind). Shouldn't we use getUnderlyingValue here as well?

We can get value back. I've done that for now (lets see what the performance implications are). 

> Shouldn't we use getUnderlyingValue here as well?

:+1:

File: javascript/ql/src/semmle/javascript/Expr.qll (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/Expr.qll:1560]: > Also see `ConcatenationRoot::getConstantStringParts`, if you haven't already.

I took heavy inspiration from there :wink: 

I also tried to share some of the implementation, but that results in non-monotonic recursion.

**Comment by erik-krogh** [comment]
> > If we don't mind that loss (I think we do mind). Shouldn't we use getUnderlyingValue here as well?
> 
> We can get value back. I've done that for now (lets see what the performance implications are).

[Performance is still tempting after the changes](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1593693741283). 

I'm doing more evaluation on nightly/big-apps over the weekend.

**Comment by erik-krogh** [comment]
[nightly/security](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1593748625306) shows a ~7% speedup on average. But some benchmarks are definitely slower.

**Comment by erik-krogh** [comment]
I did some evaluations on my laptop to get around performance wobble.  
[Here they are](https://docs.google.com/spreadsheets/d/1zh23y7MTS_tfxJGeNNmA413zOiSes9AKATpw9OkuBTg/edit#gid=0).  

[Here is a big-apps evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1594226002338). (Ignore the rightmost results, that was a failed experiment to see if I could get better performance). 

I tried to get the slowest benchmarks to perform faster, but I've had no success in this.   

This change is controversial, so I want a unanimous agreement before I'll push the merge button.

**Comment by esbena** [comment]
LGTM too.

**Pull request merged**